libc/watcom/asm/segments.asm

```diff
@@ -25,61 +25,12 @@
 ;*
 ;*  ========================================================================
 ;*
-;* Description:  Watcom C segment ordering and null pointer section offsets
+;* Description:  Watcom C segment ordering for init/fini LIBC
 ;*
 ;*****************************************************************************
 
 
-        name    cstart
-        assume  nothing
-
- DGROUP group _NULL,_AFTERNULL,CONST,STRINGS,_DATA,DATA,XIB,XI,XIE,YIB,YI,YIE,_BSS
-
-; this guarantees that no function pointer will equal NULL
-; (WLINK will keep segment 'BEGTEXT' in front)
-; This segment must be at least 2 bytes in size to avoid confusing the
-; signal function.
-; need a symbol defined here to prevent the dead code elimination from
-; eliminating the segment.
-; (the int 3h is useful for quickly revealing jumps to NULL code pointers)
-
-BEGTEXT  segment word public 'CODE'
-        assume  cs:BEGTEXT
-        int     3
-        nop
-__begtext label byte
-        assume  cs:nothing
-BEGTEXT  ends
-
-_TEXT   segment word public 'CODE'
-
-FAR_DATA segment byte public 'FAR_DATA'
-FAR_DATA ends
-
-        assume  ds:DGROUP
-
-_NULL   segment para public 'BEGDATA'
-__nullarea label word
-        db      'nul'
-        db      0
-        public  __nullarea
-_NULL   ends
-
-_AFTERNULL segment word public 'BEGDATA'
-        ;dw      0                       ; nullchar for string at address 0
-_AFTERNULL ends
-
-CONST   segment word public 'DATA'
-CONST   ends
-
-STRINGS segment word public 'DATA'
-STRINGS ends
-
-_DATA   segment word public 'DATA'
-_DATA   ends
-
-DATA    segment word public 'DATA'
-DATA    ends
```

Comment:
The reason for all these segments is that the OBJ files produced by OWC may contain references to them by name, and I thought it important to ensure that there were in fact named segments in this file in order to specify and guarantee the order they are linked. It is quite important that the compilation/linking process controls the ordering of segments, as other ia16-elf-gcc linker script magic sometimes uses them, although I have not implemented them yet. Thus, I would prefer that none of these definitions are removed, including the DGROUP above.

Also, I have some other tools (OBJ file converters, dumpers, etc) that reference these segments. These tools can be seen in elks/tools/objtools.

Reply:
The wlink use standard rules for ordering segment as is defined for OMF format that put together segments with compatible segment attributes. there is several levels of comparision. The name of segment is mostly not important except some OW special segments that change of segment name don't cause any problem. the main attribute is segment class as 'CODE' 'DATA' STACK' 'BSS' and it is also used by DOSSEG segment ordering. But change segment name has sense if you mixed modules from different toolchains because each use different names and also it depends on target system. In such case you could need some segment name changes etc. it is possible. Also it is possible to use custom segment ordering that you can setup rules for all toolchains modules without change segment names. It is not necessary to use some specialised tools it can be done by wlink configuration.

Reply:
I want to keep all named segments in segments.asm and not remove any of them. This file is the place for all segment-ordering and segment-related linker special treatment, as well as "documentation" of the application memory layout.

Rejected.